File: verification/stats.ts
/**
 * Deterministic numerical helpers: seeded PRNG, OLS with t-tests, quantiles.
 * No external dependencies so the engine remains reproducible byte-for-byte.
 */

/** mulberry32 — small, fast, deterministic. */
export function makeRng(seed: number): () => number {
  let a = seed >>> 0;
  return () => {
    a = (a + 0x6d2b79f5) >>> 0;
    let t = a;
    t = Math.imul(t ^ (t >>> 15), t | 1);
    t ^= t + Math.imul(t ^ (t >>> 7), t | 61);
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}

export function randomNormal(rng: () => number, mean = 0, sd = 1): number {
  let u = 0;
  let v = 0;
  while (u === 0) u = rng();
  while (v === 0) v = rng();
  return mean + sd * Math.sqrt(-2 * Math.log(u)) * Math.cos(2 * Math.PI * v);
}

export function mean(xs: number[]): number {
  if (xs.length === 0) return NaN;
  let s = 0;
  for (const x of xs) s += x;
  return s / xs.length;
}

export function median(xs: number[]): number {
  if (xs.length === 0) return NaN;
  const s = [...xs].sort((a, b) => a - b);
  const mid = Math.floor(s.length / 2);
  return s.length % 2 === 0 ? (s[mid - 1]! + s[mid]!) / 2 : s[mid]!;
}

export function sampleSd(xs: number[]): number {
  if (xs.length < 2) return NaN;
  const m = mean(xs);
  let s = 0;
  for (const x of xs) s += (x - m) * (x - m);
  return Math.sqrt(s / (xs.length - 1));
}

/** Linear-interpolated quantile, q in [0,1]. */
export function quantile(xs: number[], q: number): number {
  if (xs.length === 0) return NaN;
  const s = [...xs].sort((a, b) => a - b);
  const pos = (s.length - 1) * q;
  const lo = Math.floor(pos);
  const hi = Math.ceil(pos);
  if (lo === hi) return s[lo]!;
  return s[lo]! + (s[hi]! - s[lo]!) * (pos - lo);
}

export interface OlsResult {
  beta: number[];
  se: number[];
  tStat: number[];
  pValue: number[];
  n: number;
  dof: number;
  sigma2: number;
}

/**
 * Ordinary least squares via normal equations with Gaussian elimination.
 * X: n rows × p columns (row-major array of rows). Returns coefficient
 * estimates, standard errors and two-sided t-test p-values.
 */
export function ols(X: number[][], y: number[]): OlsResult {
  const n = X.length;
  const p = X[0]?.length ?? 0;
  if (n === 0 || p === 0 || y.length !== n) throw new Error('ols: bad dimensions');
  const xtx: number[][] = Array.from({ length: p }, () => Array<number>(p).fill(0));
  const xty: number[] = Array<number>(p).fill(0);
  for (let i = 0; i < n; i++) {
    const row = X[i]!;
    for (let a = 0; a < p; a++) {
      xty[a]! += row[a]! * y[i]!;
      for (let b = 0; b < p; b++) xtx[a]![b]! += row[a]! * row[b]!;
    }
  }
  const inv = invert(xtx);
  const beta = inv.map((r) => r.reduce((s, v, j) => s + v * xty[j]!, 0));
  let rss = 0;
  for (let i = 0; i < n; i++) {
    const row = X[i]!;
    let yhat = 0;
    for (let a = 0; a < p; a++) yhat += row[a]! * beta[a]!;
    rss += (y[i]! - yhat) ** 2;
  }
  const dof = n - p;
  const sigma2 = dof > 0 ? rss / dof : NaN;
  const se = beta.map((_, a) => Math.sqrt(sigma2 * inv[a]![a]!));
  const tStat = beta.map((b, a) => b / se[a]!);
  const pValue = tStat.map((t) => (dof > 0 ? 2 * (1 - studentTCdf(Math.abs(t), dof)) : NaN));
  return { beta, se, tStat, pValue, n, dof, sigma2 };
}

function invert(m: number[][]): number[][] {
  const n = m.length;
  const a = m.map((row, i) => [...row, ...Array.from({ length: n }, (_, j) => (i === j ? 1 : 0))]);
  for (let col = 0; col < n; col++) {
    let pivot = col;
    for (let r = col + 1; r < n; r++) if (Math.abs(a[r]![col]!) > Math.abs(a[pivot]![col]!)) pivot = r;
    if (Math.abs(a[pivot]![col]!) < 1e-12) throw new Error('ols: singular design matrix');
    [a[col], a[pivot]] = [a[pivot]!, a[col]!];
    const pv = a[col]![col]!;
    for (let c = 0; c < 2 * n; c++) a[col]![c]! /= pv;
    for (let r = 0; r < n; r++) {
      if (r === col) continue;
      const f = a[r]![col]!;
      if (f === 0) continue;
      for (let c = 0; c < 2 * n; c++) a[r]![c]! -= f * a[col]![c]!;
    }
  }
  return a.map((row) => row.slice(n));
}

/** Student t CDF via the regularised incomplete beta function. */
export function studentTCdf(t: number, dof: number): number {
  const x = dof / (dof + t * t);
  const ib = regularizedIncompleteBeta(x, dof / 2, 0.5);
  return t >= 0 ? 1 - 0.5 * ib : 0.5 * ib;
}

/** Regularised incomplete beta I_x(a, b) via Lentz's continued fraction (NR §6.4). */
export function regularizedIncompleteBeta(x: number, a: number, b: number): number {
  if (x <= 0) return 0;
  if (x >= 1) return 1;
  // The continued fraction converges fast only for x < (a+1)/(a+b+2); use symmetry otherwise.
  if (x > (a + 1) / (a + b + 2)) return 1 - regularizedIncompleteBeta(1 - x, b, a);
  const lbeta = lnGamma(a) + lnGamma(b) - lnGamma(a + b);
  const front = Math.exp(Math.log(x) * a + Math.log(1 - x) * b - lbeta) / a;
  const fpmin = 1e-300;
  let c = 1;
  let d = 1 - ((a + b) * x) / (a + 1);
  if (Math.abs(d) < fpmin) d = fpmin;
  d = 1 / d;
  let h = d;
  for (let m = 1; m <= 300; m++) {
    const m2 = 2 * m;
    let aa = (m * (b - m) * x) / ((a + m2 - 1) * (a + m2));
    d = 1 + aa * d;
    if (Math.abs(d) < fpmin) d = fpmin;
    c = 1 + aa / c;
    if (Math.abs(c) < fpmin) c = fpmin;
    d = 1 / d;
    h *= d * c;
    aa = (-(a + m) * (a + b + m) * x) / ((a + m2) * (a + m2 + 1));
    d = 1 + aa * d;
    if (Math.abs(d) < fpmin) d = fpmin;
    c = 1 + aa / c;
    if (Math.abs(c) < fpmin) c = fpmin;
    d = 1 / d;
    const del = d * c;
    h *= del;
    if (Math.abs(del - 1) < 3e-14) break;
  }
  return front * h;
}

function lnGamma(z: number): number {
  const g = 7;
  const coef = [
    0.99999999999980993, 676.5203681218851, -1259.1392167224028, 771.32342877765313,
    -176.61502916214059, 12.507343278686905, -0.13857109526572012, 9.9843695780195716e-6,
    1.5056327351493116e-7,
  ];
  if (z < 0.5) return Math.log(Math.PI / Math.sin(Math.PI * z)) - lnGamma(1 - z);
  z -= 1;
  let x = coef[0]!;
  for (let i = 1; i < g + 2; i++) x += coef[i]! / (z + i);
  const t = z + g + 0.5;
  return 0.5 * Math.log(2 * Math.PI) + (z + 0.5) * Math.log(t) - t + Math.log(x);
}

/** Days since 1970-01-01 for an ISO date or datetime string. */
export function dayNumber(iso: string): number {
  return Date.parse(iso.length === 10 ? `${iso}T00:00:00Z` : iso) / 86_400_000;
}

/** Fractional years since 2000-01-01, used as the OLS time axis. */
export function yearsSince2000(iso: string): number {
  return (dayNumber(iso) - dayNumber('2000-01-01')) / 365.25;
}

export function dayOfYear(iso: string): number {
  const d = new Date(iso.length === 10 ? `${iso}T00:00:00Z` : iso);
  const start = Date.UTC(d.getUTCFullYear(), 0, 1);
  return (d.getTime() - start) / 86_400_000;
}

File: analysis/ecorestore_analysis/stats.py
"""Numerical helpers mirroring ``verification/stats.ts``.

Sums are sequential (``sum`` over a Python list is left-to-right in float64),
matching the reference loops rather than numpy's pairwise summation, so the
last bit agrees. The regression and the t-distribution use numpy and scipy —
the conventional tools — and agree with the reference to well inside the six
decimals the diagnostic reports.
"""

from __future__ import annotations

import math
from dataclasses import dataclass
from datetime import datetime, timezone

import numpy as np
from scipy import stats as sps

_EPOCH = datetime(1970, 1, 1, tzinfo=timezone.utc)


def seq_sum(xs) -> float:
    """Left-to-right float64 accumulation. Python 3.12's built-in ``sum`` uses
    compensated (Neumaier) summation for floats, which is *more* accurate than
    the reference and therefore does not reproduce it."""
    s = 0.0
    for x in xs:
        s += x
    return s


def mean(xs: list[float]) -> float:
    if not xs:
        return math.nan
    return seq_sum(xs) / len(xs)


def median(xs: list[float]) -> float:
    if not xs:
        return math.nan
    s = sorted(xs)
    mid = len(s) // 2
    return (s[mid - 1] + s[mid]) / 2 if len(s) % 2 == 0 else s[mid]


def sample_sd(xs: list[float]) -> float:
    if len(xs) < 2:
        return math.nan
    m = mean(xs)
    s = 0.0
    for x in xs:
        s += (x - m) * (x - m)
    return math.sqrt(s / (len(xs) - 1))


def quantile(xs: list[float] | np.ndarray, q: float) -> float:
    """Linear-interpolated quantile, identical to the reference."""
    n = len(xs)
    if n == 0:
        return math.nan
    s = sorted(float(x) for x in xs)
    pos = (n - 1) * q
    lo = math.floor(pos)
    hi = math.ceil(pos)
    if lo == hi:
        return s[lo]
    return s[lo] + (s[hi] - s[lo]) * (pos - lo)


@dataclass
class OlsResult:
    beta: np.ndarray
    se: np.ndarray
    t_stat: np.ndarray
    p_value: np.ndarray
    n: int
    dof: int
    sigma2: float


def ols(X: np.ndarray, y: np.ndarray) -> OlsResult:
    """OLS via the normal equations with two-sided t-tests.

    Raises ``numpy.linalg.LinAlgError`` on a singular design.
    """
    n, p = X.shape
    if n == 0 or p == 0 or y.shape[0] != n:
        raise ValueError("ols: bad dimensions")
    xtx = X.T @ X
    xty = X.T @ y
    inv = np.linalg.inv(xtx)
    if not np.all(np.isfinite(inv)):
        raise np.linalg.LinAlgError("singular design matrix")
    beta = inv @ xty
    resid = y - X @ beta
    rss = float(resid @ resid)
    dof = n - p
    sigma2 = rss / dof if dof > 0 else math.nan
    se = np.sqrt(sigma2 * np.diag(inv))
    t_stat = beta / se
    p_value = 2 * sps.t.sf(np.abs(t_stat), dof) if dof > 0 else np.full(p, math.nan)
    return OlsResult(beta, se, t_stat, p_value, n, dof, sigma2)


def _epoch_ms(iso: str) -> int:
    """``Date.parse`` in milliseconds for the ISO forms the snapshot uses."""
    s = iso if len(iso) > 10 else f"{iso}T00:00:00Z"
    dt = datetime.fromisoformat(s.replace("Z", "+00:00"))
    if dt.tzinfo is None:
        dt = dt.replace(tzinfo=timezone.utc)
    td = dt - _EPOCH
    return td.days * 86_400_000 + td.seconds * 1000 + td.microseconds // 1000


def day_number(iso: str) -> float:
    return _epoch_ms(iso) / 86_400_000


_DAY_2000 = day_number("2000-01-01")


def years_since_2000(iso: str) -> float:
    return (day_number(iso) - _DAY_2000) / 365.25


def day_of_year(iso: str) -> float:
    ms = _epoch_ms(iso)
    year = datetime.fromisoformat((iso if len(iso) > 10 else f"{iso}T00:00:00Z").replace("Z", "+00:00")).year
    start = _epoch_ms(f"{year:04d}-01-01")
    return (ms - start) / 86_400_000
